### ISSUE TITLE
[Question] (HW4) TSan data race warning on weak memory

### ISSUE THREAD

**Issue by cwlo2F** [kaist-cp/cs431#871]
I'd like to ask whether it is possible that thread sanitizer wrongly detect error in weak ordering.

In HW4 concurrent hash table, I just resolved a data race warning from thread sanitizer on testing concurrent behavior of `GrowableArray`. It reported there were a read and a write at the same location.
I believe the reason was the usage of `Relaxed` ordering for CAS's failure.
I understood this by considering following scenario:

1. Threads T1 and T2 are trying to write at the same location of `x: Atomic<T>`
2. T1 successfully CAS at `x` with writing some new value in `SeqCst` ordering.
3. T2 failed at CAS at `x`, and it reads current value in `Relaxed` ordering.

I thought the thread sanitizer reports data race from step 2 and 3, as it detected the `Relaxed` read and the `SeqCst` write at the same time. I replaced every `Relaxed` in CAS into `SeqCst`, and the warning disappeared.
For comprehension, I'd like to also ask whether my explanation makes sense.

On the other hand, I think the program may work safely in our homework, because the address of allocated segments will not be updated until the end.
In other words, because each segment  is allocated exactly once (say at `x`), so I think the `Relaxed` read from CAS failure will be recovered by reading current value at `x`, and this is safe because write on `x` occurs only once.
Can we consider this argument is valid? If so, can we consider the report of TSan as false positive (or false negative)?

**Comment by Lee-Janggun** [comment]
IIRC TSAN doesn't work properly when you mix SeqCst and relaxed orderings on the same location.

**Comment by tomtomjhj** [comment]
> a read and a write at the same location.

which write/read at which location?
